File: packages/sdk-go/melaya/trade.go
package melaya

import "context"

// TradeAPI is the live, credentialed trading plane: order placement, account
// state, and position management on a CONNECTED exchange. Every method POSTs to
// /api/v1/private/<op>; the server resolves the connected key (by apiKeyId) and
// forwards to the venue via the in-house Rust engine.
//
// WARNING: the write methods (CreateOrder, CancelOrder, AmendOrder,
// CancelAllOrders, CancelPlanOrders, ClosePosition, SetLeverage, SetMarginMode,
// SetPositionMode) move REAL funds. Use the sim broker for risk-free testing.
type TradeAPI struct{ h *httpClient }

// LiveResult is the shared response envelope of every live op.
type LiveResult = map[string]interface{}

func (t *TradeAPI) op(ctx context.Context, op string, body map[string]interface{}) (LiveResult, error) {
	data, err := t.h.post(ctx, "/api/v1/private/"+op, body)
	if err != nil {
		return nil, err
	}
	var out LiveResult
	if err := unmarshal(data, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ── Account state (reads) ─────────────────────────────────────────────────

func (t *TradeAPI) Balance(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "balance", body)
}
func (t *TradeAPI) Positions(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "positions", body)
}
func (t *TradeAPI) PositionsHistory(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "positions-history", body)
}
func (t *TradeAPI) OpenOrders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "open-orders", body)
}
func (t *TradeAPI) Orders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "orders", body)
}
func (t *TradeAPI) ClosedOrders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "closed-orders", body)
}
func (t *TradeAPI) MyTrades(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "my-trades", body)
}
func (t *TradeAPI) MyTradesHistory(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "my-trades-history", body)
}
func (t *TradeAPI) PlanOrders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "plan-orders", body)
}
func (t *TradeAPI) Leverage(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "leverage", body)
}
func (t *TradeAPI) LeverageTiers(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "leverage-tiers", body)
}

// ── Order placement & management (LIVE writes — real funds) ────────────────

func (t *TradeAPI) CreateOrder(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "create-order", body)
}
func (t *TradeAPI) CancelOrder(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "cancel-order", body)
}
func (t *TradeAPI) AmendOrder(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "amend-order", body)
}
func (t *TradeAPI) CancelAllOrders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "cancel-all-orders", body)
}
func (t *TradeAPI) CancelPlanOrders(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "cancel-plan-orders", body)
}
func (t *TradeAPI) ClosePosition(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "close-position", body)
}
func (t *TradeAPI) SetLeverage(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "set-leverage", body)
}
func (t *TradeAPI) SetMarginMode(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "set-margin-mode", body)
}
func (t *TradeAPI) SetPositionMode(ctx context.Context, body map[string]interface{}) (LiveResult, error) {
	return t.op(ctx, "set-position-mode", body)
}

File: packages/sdk-rust/src/trade.rs
//! Live trading API — credentialed order placement, account state, and
//! position management on a CONNECTED exchange.
//!
//! Every method POSTs to `/api/v1/private/<op>`; the server resolves the
//! connected key (by `apiKeyId` in the body) and forwards to the venue via the
//! in-house Rust engine. Responses share the envelope
//! `{ ok, exchange, operation, orderId, clientOrderId, payload, data, ... }`.
//!
//! ⚠️ The write methods (create_order, cancel_order, amend_order,
//! cancel_all_orders, cancel_plan_orders, close_position, set_leverage,
//! set_margin_mode, set_position_mode) move REAL funds. Use the sim broker for
//! risk-free testing. Each method takes a `serde_json::Value` body (build it
//! with `serde_json::json!({ "exchange": ..., "apiKeyId": ..., ... })`).

use crate::client::HttpClient;
use crate::error::Result;
use serde_json::Value;

pub struct TradeAPI {
    http: HttpClient,
}

impl TradeAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    async fn op(&self, op: &str, body: Value) -> Result<Value> {
        self.http.post(&format!("/api/v1/private/{op}"), &body).await
    }

    // ── Account state (reads) ─────────────────────────────────────────────
    pub async fn balance(&self, body: Value) -> Result<Value> { self.op("balance", body).await }
    pub async fn positions(&self, body: Value) -> Result<Value> { self.op("positions", body).await }
    pub async fn positions_history(&self, body: Value) -> Result<Value> { self.op("positions-history", body).await }
    pub async fn open_orders(&self, body: Value) -> Result<Value> { self.op("open-orders", body).await }
    pub async fn orders(&self, body: Value) -> Result<Value> { self.op("orders", body).await }
    pub async fn closed_orders(&self, body: Value) -> Result<Value> { self.op("closed-orders", body).await }
    pub async fn my_trades(&self, body: Value) -> Result<Value> { self.op("my-trades", body).await }
    pub async fn my_trades_history(&self, body: Value) -> Result<Value> { self.op("my-trades-history", body).await }
    pub async fn plan_orders(&self, body: Value) -> Result<Value> { self.op("plan-orders", body).await }
    pub async fn leverage(&self, body: Value) -> Result<Value> { self.op("leverage", body).await }
    pub async fn leverage_tiers(&self, body: Value) -> Result<Value> { self.op("leverage-tiers", body).await }

    // ── Order placement & management (LIVE writes — real funds) ───────────
    pub async fn create_order(&self, body: Value) -> Result<Value> { self.op("create-order", body).await }
    pub async fn cancel_order(&self, body: Value) -> Result<Value> { self.op("cancel-order", body).await }
    pub async fn amend_order(&self, body: Value) -> Result<Value> { self.op("amend-order", body).await }
    pub async fn cancel_all_orders(&self, body: Value) -> Result<Value> { self.op("cancel-all-orders", body).await }
    pub async fn cancel_plan_orders(&self, body: Value) -> Result<Value> { self.op("cancel-plan-orders", body).await }
    pub async fn close_position(&self, body: Value) -> Result<Value> { self.op("close-position", body).await }
    pub async fn set_leverage(&self, body: Value) -> Result<Value> { self.op("set-leverage", body).await }
    pub async fn set_margin_mode(&self, body: Value) -> Result<Value> { self.op("set-margin-mode", body).await }
    pub async fn set_position_mode(&self, body: Value) -> Result<Value> { self.op("set-position-mode", body).await }
}
